### ISSUE TITLE
3.5.7 No more result from API when `workflows.argoproj.io/phase=Running`

### ISSUE THREAD

**Issue by hmoulart** [argoproj/argo-workflows#13143]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

Hello,
We have likely identified a regression in version 3.5.7. This is unexpected, as there are no related issues or changes noted in the changelog. The Argo Workflows API does not return any results when queried with `workflows.argoproj.io/phase=Running`, `workflows.argoproj.io/phase=Pending`. `Failed`, `Succeeded`, `Error` are OK.

```
GET https://[URL]/api/v1/workflows/[NAMESPACE]?listOptions.labelSelector=workflows.argoproj.io/phase=Running&listOptions.limit=10
```

Version 3.5.7:
![image](https://github.com/argoproj/argo-workflows/assets/16108020/4a005cf5-cb49-47d1-b13a-7b9c95a3277e)

Version 3.5.6:
![image](https://github.com/argoproj/argo-workflows/assets/16108020/923c24f1-a29f-42c8-8cdb-77fa3860ac09)

Thank you for your help on that issue
Hugo

### Version

3.5.7

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
GET https://[URL]/api/v1/workflows/[NAMESPACE]?listOptions.labelSelector=workflows.argoproj.io/phase=Running&listOptions.limit=10
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

**Comment by Joibel** [comment]
I imagine this is due to #13021, I can't see another likely source of the problem.

**Comment by agilgur5** [comment]
Yes it would be from there, cc @jiachengxu. That would suggest the archived merging logic has a bug in it -- archived workflows will never be in `Pending` state so it should only retrieve the live ones in this case 🤔

**Comment by jiachengxu** [comment]
I tried to reproduce the issue locally with v3.5.7, but I couldn't reproduce the issue, the followings are screenshots from my test, and both UI and postman returned the running workflows.
<img width="1610" alt="image" src="https://github.com/argoproj/argo-workflows/assets/18415314/0a9f76a6-fb9b-487b-b93a-7810f6caaee2">
<img width="1493" alt="image" src="https://github.com/argoproj/argo-workflows/assets/18415314/a8f57a35-9984-498f-a506-e9ddc96e48e4">

@Joibel IIRC, you are also running v3.5.7 in some of your environments, have you seen this issue before?

**Comment by hmoulart** [comment]
Probably important information, we are using persistance with PostgreSQL database. The behaviour could be different on that case?

**Comment by jiachengxu** [comment]
> Probably important information, we are using persistance with PostgreSQL database. The behaviour could be different on that case?

@hmoulart Thanks for providing the info, during my test, I was also using the postgres database for archiving workflows

**Comment by agilgur5** [comment]
Thanks for trying to repro Alan and Jiacheng!

@hmoulart I imagine there must be some other confounding variable here, perhaps in your Argo configuration, if it hasn't been easily reproduced

**Comment by tooptoop4** [comment]
@hmoulart have u tried v3.5.8?

**Comment by hmoulart** [comment]
@tooptoop4 I'm going to try tomorrow :) Thanks

**Comment by hmoulart** [comment]
Hello,
I upgraded to version 3.5.8, and I’m pleased to report that the bug has been resolved!
Thank you to suggest me this upgrade.
Hugo

**Comment by agilgur5** [comment]
For reference, that means this was resolved by #13166
